// nodebb-plugin-moti-api/library.js
'use strict';

const express = require('express');
const crypto = require('crypto');

const db = require.main.require('./src/database');
const user = require.main.require('./src/user');
const topics = require.main.require('./src/topics');
const posts = require.main.require('./src/posts');
const categories = require.main.require('./src/categories');
const search = require.main.require('./src/search');

const wechat = require('./lib/wechat');
const payment = require('./lib/payment');

// ============ 工具函数 ============

function ok(data) {
  return { success: true, data };
}

function err(message) {
  return { success: false, error: message };
}

// ============ 中间件 ============

/**
 * Token 认证中间件 - 支持小程序 Bearer Token 和 Web 端 Session
 */
async function authenticateMotiToken(req, res, next) {
  // 首先检查 Web 端的 Session 登录（NodeBB 原生）
  if (req.uid && parseInt(req.uid, 10) > 0) {
    req.loggedIn = true;
    next();
    return;
  }

  // 然后检查小程序的 Bearer Token
  const authHeader = req.headers.authorization;

  if (!authHeader || !authHeader.startsWith('Bearer ')) {
    req.uid = 0;
    req.loggedIn = false;
    next();
    return;
  }

  const token = authHeader.substring(7);
  const data = await wechat.verifyToken(token);

  if (!data) {
    req.uid = 0;
    req.loggedIn = false;
  } else {
    req.uid = data.uid;
    req.loggedIn = true;
    req.motiToken = token;
    req.openid = data.openid;
  }

  next();
}

/**
 * 要求登录中间件
 */
function requireLogin(req, res, next) {
  if (!req.loggedIn || parseInt(req.uid, 10) <= 0) {
    res.status(401).json(err('not-logged-in'));
    return;
  }
  next();
}

/**
 * 阅读权限检查（VIP 或每日限制）
 */
async function checkReadPermission(req, res, next) {
  const uid = parseInt(req.uid, 10) || 0;
  // 游客模式：暂时允许未登录用户访问
  if (uid <= 0) {
    next();
    return;
  }

  // 检查 VIP 状态
  const vipExpire = await db.getObjectField(`user:${uid}`, 'vip_expire_time');
  if (vipExpire && parseInt(vipExpire, 10) > Date.now()) {
    next();
    return;
  }

  // 非 VIP 用户检查每日阅读限制
  const today = new Date().toDateString();
  const lastRead = await db.getObjectField(`user:${uid}`, 'last_read_date');
  let readCount = await db.getObjectField(`user:${uid}`, 'daily_read_count');
  readCount = parseInt(readCount, 10) || 0;

  if (lastRead !== today) {
    await db.setObject(`user:${uid}`, { last_read_date: today, daily_read_count: 0 });
    readCount = 0;
  }

  if (readCount >= 1) {
    res.status(403).json(err('daily-limit-exceeded'));
    return;
  }

  await db.incrObjectField(`user:${uid}`, 'daily_read_count');
  next();
}

// ============ 辅助函数 ============

async function getPagination(req) {
  const page = parseInt(req.query.page, 10) || 1;
  const limit = parseInt(req.query.limit, 10) || 20;
  const start = (page - 1) * limit;
  const stop = start + limit - 1;
  return { page, limit, start, stop };
}

function mapSort(sort) {
  if (sort === 'popular') return 'votes';
  return 'recent';
}

async function getUserExtended(uid) {
  const base = await user.getUserFields(uid, ['uid', 'username', 'userslug', 'picture', 'fullname']);
  const ext = await db.getObject(`user:${uid}`) || {};
  const isVip = ext.vip_expire_time && parseInt(ext.vip_expire_time, 10) > Date.now();
  return {
    ...base,
    isVip: !!isVip,
    points: parseInt(ext.points || 0, 10) || 0,
    vipExpire: parseInt(ext.vip_expire_time || 0, 10) || 0,
    dailyReads: parseInt(ext.daily_read_count || 0, 10) || 0,
    phone: ext.phone || null,
  };
}

// ============ 笔记功能 ============

/**
 * 创建划线笔记
 * @param {number} uid - 用户ID
 * @param {object} body - { pid, content, note, startOffset, endOffset, color, visibility }
 */
async function createNote(uid, body) {
  const noteId = crypto.randomUUID();
  const now = Date.now();
  const pid = parseInt(body.pid, 10);
  const startOffset = parseInt(body.startOffset || 0, 10) || 0;
  const endOffset = parseInt(body.endOffset || 0, 10) || 0;
  const visibility = body.visibility === 'public' ? 'public' : 'private';

  const data = {
    noteId,
    uid,
    pid,
    content: String(body.content || ''),
    note: String(body.note || ''),
    startOffset,
    endOffset,
    color: String(body.color || '#ffeb3b'),
    timestamp: now,
    visibility,
    likeCount: 0,
  };

  await db.setObject(`note:${noteId}`, data);

  // 用户的所有划线索引
  await db.sortedSetAdd(`user:${uid}:notes`, now, noteId);
  // 用户在该帖子的划线索引
  await db.sortedSetAdd(`post:${pid}:user:${uid}:notes`, now, noteId);

  // 如果是公开划线，添加公开索引和聚合
  if (visibility === 'public') {
    await addPublicNoteIndexes(noteId, pid, startOffset, endOffset, now);
  }

  // 创建笔记获得积分 (+3)
  const POINTS_FOR_NOTE = 3;
  await db.incrObjectFieldBy(`user:${uid}`, 'moti_points', POINTS_FOR_NOTE);
  const totalPoints = await db.getObjectField(`user:${uid}`, 'moti_points') || 0;
  await db.sortedSetAdd('moti:points:ranking', totalPoints, uid);

  const logId = `${uid}:${now}:${Math.random().toString(36).substr(2, 6)}`;
  await db.setObject(`moti:points:log:${logId}`, {
    logId, uid, points: POINTS_FOR_NOTE, reason: 'create_note', createTime: now, noteId,
  });
  await db.sortedSetAdd(`user:${uid}:points:log`, now, logId);

  return data;
}

/**
 * 添加公开划线索引
 */
async function addPublicNoteIndexes(noteId, pid, startOffset, endOffset, timestamp) {
  // 帖子的公开划线索引
  await db.sortedSetAdd(`post:${pid}:notes:public`, timestamp, noteId);

  // 聚合位置索引
  const highlightKey = `post:${pid}:highlight:${startOffset}:${endOffset}`;
  await db.setAdd(highlightKey, noteId);

  // 更新位置的划线人数
  const count = await db.setCount(highlightKey);
  await db.sortedSetAdd(`post:${pid}:highlights`, count, `${startOffset}:${endOffset}`);
}

/**
 * 移除公开划线索引
 */
async function removePublicNoteIndexes(noteId, pid, startOffset, endOffset) {
  // 移除公开索引
  await db.sortedSetRemove(`post:${pid}:notes:public`, noteId);

  // 移除聚合位置索引
  const highlightKey = `post:${pid}:highlight:${startOffset}:${endOffset}`;
  await db.setRemove(highlightKey, noteId);

  // 更新位置的划线人数
  const count = await db.setCount(highlightKey);
  if (count > 0) {
    await db.sortedSetAdd(`post:${pid}:highlights`, count, `${startOffset}:${endOffset}`);
  } else {
    await db.sortedSetRemove(`post:${pid}:highlights`, `${startOffset}:${endOffset}`);
  }
}

/**
 * 更新划线可见性
 */
async function updateNoteVisibility(noteId, newVisibility) {
  const note = await db.getObject(`note:${noteId}`);
  if (!note) return null;

  const oldVisibility = note.visibility || 'private';
  if (oldVisibility === newVisibility) return note;

  await db.setObjectField(`note:${noteId}`, 'visibility', newVisibility);

  const { pid, startOffset, endOffset, timestamp } = note;

  if (newVisibility === 'public') {
    await addPublicNoteIndexes(noteId, pid, startOffset, endOffset, timestamp);
  } else {
    await removePublicNoteIndexes(noteId, pid, startOffset, endOffset);
  }

  return { ...note, visibility: newVisibility };
}

/**
 * 获取划线列表（带用户信息）
 */
async function getNotesBySet(set, start, stop) {
  const ids = await db.getSortedSetRevRange(set, start, stop);
  if (!ids.length) return [];

  const keys = ids.map(id => `note:${id}`);
  const notes = await db.getObjects(keys);
  const validNotes = notes.filter(Boolean);

  // 获取用户信息
  const uids = [...new Set(validNotes.map(n => n.uid))];
  const users = await user.getUsersFields(uids, ['uid', 'username', 'picture']);
  const userMap = {};
  users.forEach(u => { userMap[u.uid] = u; });

  return validNotes.map(n => ({
    ...n,
    user: userMap[n.uid] || null,
  }));
}

/**
 * 获取帖子的公开划线
 */
async function getPublicNotesByPost(pid, start, stop) {
  return getNotesBySet(`post:${pid}:notes:public`, start, stop);
}

/**
 * 获取帖子的聚合位置列表
 */
async function getHighlightsByPost(pid, start, stop) {
  const positions = await db.getSortedSetRevRange(`post:${pid}:highlights`, start, stop);
  const results = [];

  for (const pos of positions) {
    const [startOffset, endOffset] = pos.split(':').map(Number);
    const highlightKey = `post:${pid}:highlight:${startOffset}:${endOffset}`;
    const noteIds = await db.getSetMembers(highlightKey);
    const count = noteIds.length;

    // 获取第一个划线的内容作为预览
    let preview = '';
    if (noteIds.length > 0) {
      const firstNote = await db.getObject(`note:${noteIds[0]}`);
      if (firstNote) preview = firstNote.content;
    }

    results.push({
      startOffset,
      endOffset,
      count,
      preview,
      noteIds,
    });
  }

  return results;
}

/**
 * 获取某位置的所有划线
 */
async function getNotesByPosition(pid, startOffset, endOffset) {
  const highlightKey = `post:${pid}:highlight:${startOffset}:${endOffset}`;
  const noteIds = await db.getSetMembers(highlightKey);
  if (!noteIds.length) return [];

  const keys = noteIds.map(id => `note:${id}`);
  const notes = await db.getObjects(keys);
  const validNotes = notes.filter(Boolean);

  // 获取用户信息
  const uids = [...new Set(validNotes.map(n => n.uid))];
  const users = await user.getUsersFields(uids, ['uid', 'username', 'picture']);
  const userMap = {};
  users.forEach(u => { userMap[u.uid] = u; });

  return validNotes.map(n => ({
    ...n,
    user: userMap[n.uid] || null,
  }));
}

/**
 * 点赞划线
 */
async function likeNote(noteId, uid) {
  const note = await db.getObject(`note:${noteId}`);
  if (!note) return null;

  const likeKey = `note:${noteId}:likes`;
  const isLiked = await db.isSetMember(likeKey, uid);

  if (isLiked) {
    return { liked: true, likeCount: parseInt(note.likeCount, 10) || 0 };
  }

  await db.setAdd(likeKey, uid);
  await db.incrObjectField(`note:${noteId}`, 'likeCount');
  const newCount = (parseInt(note.likeCount, 10) || 0) + 1;

  return { liked: true, likeCount: newCount };
}

/**
 * 取消点赞划线
 */
async function unlikeNote(noteId, uid) {
  const note = await db.getObject(`note:${noteId}`);
  if (!note) return null;

  const likeKey = `note:${noteId}:likes`;
  const isLiked = await db.isSetMember(likeKey, uid);

  if (!isLiked) {
    return { liked: false, likeCount: parseInt(note.likeCount, 10) || 0 };
  }

  await db.setRemove(likeKey, uid);
  await db.decrObjectField(`note:${noteId}`, 'likeCount');
  const newCount = Math.max(0, (parseInt(note.likeCount, 10) || 0) - 1);

  return { liked: false, likeCount: newCount };
}

/**
 * 检查是否已点赞
 */
async function hasLikedNote(noteId, uid) {
  return db.isSetMember(`note:${noteId}:likes`, uid);
}

/**
 * 删除划线
 */
async function deleteNote(noteId, uid) {
  const note = await db.getObject(`note:${noteId}`);
  if (!note) return false;
  if (String(note.uid) !== String(uid)) return false;

  const { pid, startOffset, endOffset, visibility } = note;

  // 移除用户索引
  await db.sortedSetRemove(`user:${uid}:notes`, noteId);
  await db.sortedSetRemove(`post:${pid}:user:${uid}:notes`, noteId);

  // 如果是公开的，移除公开索引
  if (visibility === 'public') {
    await removePublicNoteIndexes(noteId, pid, startOffset, endOffset);
  }

  // 移除点赞数据
  await db.delete(`note:${noteId}:likes`);

  // 删除划线本身
  await db.delete(`note:${noteId}`);

  return true;
}

async function ensureNoteOwner(noteId, uid) {
  const owner = await db.getObjectField(`note:${noteId}`, 'uid');
  return String(owner) === String(uid);
}

// ============ 插件初始化 ============

const plugin = {};

plugin.init = async (params) => {
  const { app, middleware } = params;
  const api = express.Router();

  // CORS 跨域支持
  api.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, OPTIONS');
    res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization');
    if (req.method === 'OPTIONS') {
      return res.sendStatus(200);
    }
    next();
  });

  // 使用 NodeBB 的用户中间件来获取 Web 端的登录状态
  api.use(middleware.authenticateRequest);
  api.use(middleware.ensureLoggedIn ? (req, res, next) => {
    // 不强制登录，只是获取用户信息
    next();
  } : (req, res, next) => next());

  app.use('/api/moti', api);

  // ============ 认证相关路由 ============

  // 测试接口
  api.get('/test', (req, res) => {
    res.json(ok({ message: '莫提之地API插件运行正常', version: '2.0.0', timestamp: Date.now() }));
  });

  /**
   * 微信登录
   * POST /api/moti/auth/wechat
   * Body: { code: string, userInfo?: { nickName, avatarUrl } }
   */
  api.post('/auth/wechat', async (req, res) => {
    try {
      const { code, userInfo } = req.body || {};

      if (!code) {
        res.status(400).json(err('code-required'));
        return;
      }

      // 1. 调用微信 code2Session
      const wechatData = await wechat.code2Session(code);

      // 2. 查找已绑定用户
      let uid = await wechat.findUserByOpenid(wechatData.openid);
      let isNewUser = false;

      if (uid) {
        // 更新 session_key
        await wechat.updateSessionKey(uid, wechatData.sessionKey);
      } else {
        // 3. 创建新用户
        uid = await wechat.createWechatUser(wechatData, userInfo || {});
        isNewUser = true;
      }

      // 4. 生成 Token
      const tokenData = await wechat.generateToken(uid, wechatData.openid);

      // 5. 获取用户信息
      const userData = await getUserExtended(uid);

      res.json(ok({
        token: tokenData.token,
        expireTime: tokenData.expireTime,
        user: userData,
        isNewUser,
      }));
    } catch (e) {
      console.error('[moti-api] wechat login error:', e);
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 用户名密码登录
   * POST /api/moti/auth/login
   * Body: { username: string, password: string }
   */
  api.post('/auth/login', async (req, res) => {
    try {
      const { username, password } = req.body || {};

      if (!username || !password) {
        res.status(400).json(err('用户名和密码不能为空'));
        return;
      }

      // 验证用户名密码
      const uid = await user.getUidByUserslug(username.toLowerCase());
      if (!uid) {
        res.status(401).json(err('用户名或密码错误'));
        return;
      }

      const isValid = await user.isPasswordCorrect(uid, password, req.ip);
      if (!isValid) {
        res.status(401).json(err('用户名或密码错误'));
        return;
      }

      // 生成 Token
      const token = crypto.randomBytes(32).toString('hex');
      const expireTime = Date.now() + 30 * 24 * 60 * 60 * 1000; // 30天

      await db.setObject(`moti:token:${token}`, {
        uid: uid,
        createTime: Date.now(),
        expireTime: expireTime
      });
      await db.sortedSetAdd(`moti:user:${uid}:tokens`, Date.now(), token);

      // 获取用户信息
      const userData = await getUserExtended(uid);

      res.json(ok({
        token: token,
        expireTime: expireTime,
        user: userData
      }));
    } catch (e) {
      console.error('[moti-api] login error:', e);
      res.status(500).json(err('登录失败'));
    }
  });

  /**
   * 刷新 Token
   * POST /api/moti/auth/refresh
   * Header: Authorization: Bearer <token>
   */
  api.post('/auth/refresh', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const newTokenData = await wechat.refreshToken(req.motiToken);

      if (!newTokenData) {
        res.status(401).json(err('token-invalid'));
        return;
      }

      res.json(ok({
        token: newTokenData.token,
        expireTime: newTokenData.expireTime,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 退出登录
   * POST /api/moti/auth/logout
   */
  api.post('/auth/logout', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      await wechat.revokeToken(req.motiToken);
      res.json(ok({ message: 'logged-out' }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 绑定手机号
   * POST /api/moti/auth/bindPhone
   * Body: { encryptedData: string, iv: string } 或 { phone: string }（测试用）
   */
  api.post('/auth/bindPhone', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { encryptedData, iv, phone } = req.body || {};
      let phoneNumber = phone;

      // 如果提供了加密数据，则解密获取手机号
      if (encryptedData && iv) {
        const wechatBinding = await wechat.getWechatBinding(req.uid);
        if (!wechatBinding || !wechatBinding.sessionKey) {
          res.status(400).json(err('session-expired'));
          return;
        }

        const decrypted = wechat.decryptWechatData(wechatBinding.sessionKey, encryptedData, iv);
        phoneNumber = decrypted.phoneNumber || decrypted.purePhoneNumber;
      }

      if (!phoneNumber) {
        res.status(400).json(err('phone-required'));
        return;
      }

      await wechat.bindPhone(req.uid, phoneNumber);
      const userData = await getUserExtended(req.uid);

      res.json(ok({ user: userData }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 用户相关路由 ============

  /**
   * 获取当前用户资料
   * GET /api/moti/user/profile
   */
  api.get('/user/profile', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const data = await getUserExtended(req.uid);
      res.json(ok(data));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 更新用户资料
   * PUT /api/moti/user/profile
   * Body: { nickName?: string, avatarUrl?: string }
   */
  api.put('/user/profile', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { nickName, avatarUrl } = req.body || {};
      const updates = {};

      if (nickName !== undefined) {
        updates.fullname = String(nickName).substring(0, 50);
      }
      if (avatarUrl !== undefined) {
        updates.picture = String(avatarUrl);
      }

      if (Object.keys(updates).length > 0) {
        await user.updateProfile(req.uid, updates, [req.uid]);
      }

      const data = await getUserExtended(req.uid);
      res.json(ok(data));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 内容浏览路由 ============

  /**
   * 获取帖子列表
   * GET /api/moti/posts
   * Query: page, limit, sort (recent|popular)
   */
  api.get('/posts', [authenticateMotiToken], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);
      const sort = mapSort(String(req.query.sort || 'recent'));
      const result = await topics.getSortedTopics({ uid: req.uid || 0, start, stop, sort });
      res.json(ok({
        topics: result.topics,
        page,
        pageCount: result.nextStart > 0 ? Math.ceil((result.topicCount || 0) / limit) : 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取帖子详情
   * GET /api/moti/post/:pid
   */
  api.get('/post/:pid', [authenticateMotiToken, checkReadPermission], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      const topicData = await topics.getTopicDataByPid(pid);
      const tid = topicData && topicData.tid;
      const set = `tid:${tid}:posts`;
      const postsData = await topics.getTopicPosts(topicData, set, 0, 20, req.uid, false);

      // 标记为已读
      if (req.uid && tid) {
        await topics.markAsRead([tid], req.uid);
      }

      res.json(ok({ topic: topicData, posts: postsData }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 搜索帖子
   * GET /api/moti/search
   * Query: q (关键词), page, limit, cid (分类ID), tags (标签,逗号分隔), sort (relevance|timestamp)
   */
  api.get('/search', [authenticateMotiToken], async (req, res) => {
    try {
      const { page, limit } = await getPagination(req);
      const query = String(req.query.q || '').trim();

      if (!query) {
        res.status(400).json(err('query-required'));
        return;
      }

      const searchData = {
        query,
        searchIn: 'titlesposts',
        uid: req.uid || 0,
        page,
        itemsPerPage: limit,
        sortBy: req.query.sort || 'relevance',
      };

      // 分类筛选
      if (req.query.cid) {
        searchData.categories = [parseInt(req.query.cid, 10)];
      }

      // 标签筛选
      if (req.query.tags) {
        searchData.hasTags = String(req.query.tags).split(',').map(t => t.trim()).filter(Boolean);
      }

      const result = await search.search(searchData);
      res.json(ok({
        posts: result.posts || [],
        matchCount: result.matchCount || 0,
        page,
        pageCount: result.pageCount || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取分类列表
   * GET /api/moti/categories
   */
  api.get('/categories', [authenticateMotiToken], async (req, res) => {
    try {
      const cids = await categories.getAllCidsFromSet('categories:cid');
      const categoryData = await categories.getCategoriesData(cids);
      const filtered = categoryData.filter(c => c && !c.disabled);
      res.json(ok({ categories: filtered }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取分类下的帖子
   * GET /api/moti/category/:cid/topics
   * Query: page, limit, sort (recent|popular|posts|views)
   */
  api.get('/category/:cid/topics', [authenticateMotiToken], async (req, res) => {
    try {
      const cid = parseInt(req.params.cid, 10);
      const { page, limit, start, stop } = await getPagination(req);

      // 排序映射
      const sortMap = {
        recent: 'recently_created',
        popular: 'most_votes',
        posts: 'most_posts',
        views: 'most_views',
        replied: 'recently_replied',
      };
      const sort = sortMap[req.query.sort] || 'recently_replied';

      const result = await categories.getCategoryTopics({
        cid,
        uid: req.uid || 0,
        start,
        stop,
        sort,
      });

      res.json(ok({
        topics: result.topics || [],
        page,
        pageCount: Math.ceil((result.topic_count || 0) / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取标签列表
   * GET /api/moti/tags
   */
  api.get('/tags', [authenticateMotiToken], async (req, res) => {
    try {
      const tags = await topics.getTags(0, 99);
      res.json(ok({ tags }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取指定标签的帖子
   * GET /api/moti/tag/:tag/topics
   */
  api.get('/tag/:tag/topics', [authenticateMotiToken], async (req, res) => {
    try {
      const tag = String(req.params.tag);
      const { page, limit, start, stop } = await getPagination(req);
      const result = await topics.getTagTids(tag, start, stop);
      const tids = result || [];
      const topicsData = await topics.getTopics(tids, req.uid || 0);
      res.json(ok({
        topics: topicsData.filter(Boolean),
        page,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取未读帖子列表
   * GET /api/moti/unread
   * Query: page, limit, filter (空|new|watched|unreplied)
   */
  api.get('/unread', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);
      const filter = String(req.query.filter || '');

      const result = await topics.getUnreadTopics({
        uid: req.uid,
        start,
        stop,
        filter,
      });

      res.json(ok({
        topics: result.topics || [],
        unreadCount: result.topicCount || 0,
        page,
        pageCount: Math.ceil((result.topicCount || 0) / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 标记帖子为已读
   * POST /api/moti/topics/read
   * Body: { tids: [tid1, tid2, ...] } 或 { all: true }
   */
  api.post('/topics/read', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { tids, all } = req.body || {};

      if (all) {
        await topics.markAllRead(req.uid);
      } else if (tids && Array.isArray(tids) && tids.length > 0) {
        await topics.markAsRead(tids.map(t => parseInt(t, 10)), req.uid);
      } else {
        res.status(400).json(err('tids-or-all-required'));
        return;
      }

      res.json(ok({ marked: true }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 互动功能路由 ============

  /**
   * 点赞帖子
   * POST /api/moti/post/:pid/upvote
   */
  api.post('/post/:pid/upvote', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      const r = await posts.upvote(pid, req.uid);
      res.json(ok({ upvote: r.upvote, votes: r.post ? r.post.votes : undefined }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 取消点赞
   * DELETE /api/moti/post/:pid/upvote
   */
  api.delete('/post/:pid/upvote', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      const r = await posts.unvote(pid, req.uid);
      res.json(ok({ votes: r.post ? r.post.votes : undefined }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 收藏帖子
   * POST /api/moti/post/:pid/bookmark
   */
  api.post('/post/:pid/bookmark', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      await posts.bookmark(pid, req.uid);
      res.json(ok({ bookmarked: true }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 取消收藏
   * DELETE /api/moti/post/:pid/bookmark
   */
  api.delete('/post/:pid/bookmark', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      await posts.unbookmark(pid, req.uid);
      res.json(ok({ bookmarked: false }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 发表评论
   * POST /api/moti/post/:pid/comment
   * Body: { content: string, toPid?: number }
   */
  api.post('/post/:pid/comment', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      if (!req.body || !req.body.content) {
        res.status(400).json(err('content-required'));
        return;
      }
      const pid = parseInt(req.params.pid, 10);
      const tid = await posts.getPostField(pid, 'tid');
      const data = await posts.create({
        uid: req.uid,
        tid,
        content: String(req.body.content),
        toPid: req.body.toPid ? parseInt(req.body.toPid, 10) : undefined,
      });
      res.json(ok(data));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取评论列表
   * GET /api/moti/post/:pid/comments
   */
  api.get('/post/:pid/comments', [authenticateMotiToken], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      const tid = await posts.getPostField(pid, 'tid');
      const topicData = await topics.getTopicData(tid);
      const { start, stop } = await getPagination(req);
      const repliesStart = start === 0 ? 1 : start;
      const list = await topics.getTopicPosts(topicData, `tid:${tid}:posts`, repliesStart, stop, req.uid || 0, false);
      res.json(ok({ comments: list }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 划线笔记路由 ============

  /**
   * 创建笔记
   * POST /api/moti/notes
   * Body: { pid, content, note?, startOffset, endOffset, color?, visibility? }
   */
  api.post('/notes', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      if (!req.body || !req.body.pid) {
        res.status(400).json(err('pid-required'));
        return;
      }
      const note = await createNote(req.uid, req.body);
      res.json(ok(note));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取我的笔记列表
   * GET /api/moti/notes
   */
  api.get('/notes', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { start, stop } = await getPagination(req);
      const notes = await getNotesBySet(`user:${req.uid}:notes`, start, stop);
      res.json(ok({ notes }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取我在某帖子的笔记
   * GET /api/moti/post/:pid/notes/mine
   */
  api.get('/post/:pid/notes/mine', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { start, stop } = await getPagination(req);
      const pid = parseInt(req.params.pid, 10);
      const notes = await getNotesBySet(`post:${pid}:user:${req.uid}:notes`, start, stop);
      res.json(ok({ notes }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取帖子的公开划线列表
   * GET /api/moti/post/:pid/notes/public
   */
  api.get('/post/:pid/notes/public', [authenticateMotiToken], async (req, res) => {
    try {
      const { start, stop } = await getPagination(req);
      const pid = parseInt(req.params.pid, 10);
      const notes = await getPublicNotesByPost(pid, start, stop);

      // 如果用户已登录，标记是否已点赞
      if (req.uid) {
        for (const note of notes) {
          note.liked = await hasLikedNote(note.noteId, req.uid);
        }
      }

      res.json(ok({ notes }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取帖子的聚合划线位置（热门位置）
   * GET /api/moti/post/:pid/highlights
   */
  api.get('/post/:pid/highlights', [authenticateMotiToken], async (req, res) => {
    try {
      const { start, stop } = await getPagination(req);
      const pid = parseInt(req.params.pid, 10);
      const highlights = await getHighlightsByPost(pid, start, stop);
      res.json(ok({ highlights }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取某位置的所有划线
   * GET /api/moti/post/:pid/highlight/:start/:end
   */
  api.get('/post/:pid/highlight/:start/:end', [authenticateMotiToken], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      const startOffset = parseInt(req.params.start, 10);
      const endOffset = parseInt(req.params.end, 10);
      const notes = await getNotesByPosition(pid, startOffset, endOffset);

      // 如果用户已登录，标记是否已点赞
      if (req.uid) {
        for (const note of notes) {
          note.liked = await hasLikedNote(note.noteId, req.uid);
        }
      }

      res.json(ok({ notes }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 更新笔记
   * PUT /api/moti/notes/:noteId
   * Body: { note?, color?, visibility? }
   */
  api.put('/notes/:noteId', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const noteId = String(req.params.noteId);
      const owner = await ensureNoteOwner(noteId, req.uid);
      if (!owner) {
        res.status(403).json(err('not-allowed'));
        return;
      }

      // 处理 visibility 更新
      if (req.body && req.body.visibility !== undefined) {
        const newVisibility = req.body.visibility === 'public' ? 'public' : 'private';
        await updateNoteVisibility(noteId, newVisibility);
      }

      // 处理其他字段更新
      const updates = {};
      if (req.body && req.body.note !== undefined) updates.note = String(req.body.note);
      if (req.body && req.body.color !== undefined) updates.color = String(req.body.color);
      if (Object.keys(updates).length) {
        await db.setObject(`note:${noteId}`, updates);
      }

      const data = await db.getObject(`note:${noteId}`);
      res.json(ok(data));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 删除笔记
   * DELETE /api/moti/notes/:noteId
   */
  api.delete('/notes/:noteId', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const noteId = String(req.params.noteId);
      const owner = await ensureNoteOwner(noteId, req.uid);
      if (!owner) {
        res.status(403).json(err('not-allowed'));
        return;
      }
      const deleted = await deleteNote(noteId, req.uid);
      res.json(ok({ deleted }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 点赞划线
   * POST /api/moti/notes/:noteId/like
   */
  api.post('/notes/:noteId/like', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const noteId = String(req.params.noteId);
      const result = await likeNote(noteId, req.uid);
      if (!result) {
        res.status(404).json(err('note-not-found'));
        return;
      }
      res.json(ok(result));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 取消点赞划线
   * DELETE /api/moti/notes/:noteId/like
   */
  api.delete('/notes/:noteId/like', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const noteId = String(req.params.noteId);
      const result = await unlikeNote(noteId, req.uid);
      if (!result) {
        res.status(404).json(err('note-not-found'));
        return;
      }
      res.json(ok(result));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 用户记录路由 ============

  /**
   * 获取我的收藏列表
   * GET /api/moti/user/bookmarks
   * Query: page, limit
   */
  api.get('/user/bookmarks', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);

      // 获取收藏的帖子 ID
      const pids = await db.getSortedSetRevRange(`uid:${req.uid}:bookmarks`, start, stop);
      const total = await db.sortedSetCard(`uid:${req.uid}:bookmarks`);

      if (!pids.length) {
        res.json(ok({ posts: [], page, pageCount: 1 }));
        return;
      }

      // 获取帖子数据
      const postsData = await posts.getPostSummaryByPids(pids, req.uid, { stripTags: true });

      res.json(ok({
        posts: postsData,
        page,
        pageCount: Math.ceil(total / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取我的点赞记录
   * GET /api/moti/user/upvotes
   */
  api.get('/user/upvotes', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);

      const pids = await db.getSortedSetRevRange(`uid:${req.uid}:upvote`, start, stop);
      const total = await db.sortedSetCard(`uid:${req.uid}:upvote`);

      if (!pids.length) {
        res.json(ok({ posts: [], page, pageCount: 1 }));
        return;
      }

      const postsData = await posts.getPostSummaryByPids(pids, req.uid, { stripTags: true });

      res.json(ok({
        posts: postsData,
        page,
        pageCount: Math.ceil(total / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取我发表的帖子/评论
   * GET /api/moti/user/posts
   */
  api.get('/user/posts', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);

      const pids = await db.getSortedSetRevRange(`uid:${req.uid}:posts`, start, stop);
      const total = await db.sortedSetCard(`uid:${req.uid}:posts`);

      if (!pids.length) {
        res.json(ok({ posts: [], page, pageCount: 1 }));
        return;
      }

      const postsData = await posts.getPostSummaryByPids(pids, req.uid, { stripTags: true });

      res.json(ok({
        posts: postsData,
        page,
        pageCount: Math.ceil(total / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取别人对我的点赞（被点赞记录）
   * GET /api/moti/user/received-upvotes
   */
  api.get('/user/received-upvotes', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);

      // 获取用户发表的帖子
      const myPids = await db.getSortedSetRevRange(`uid:${req.uid}:posts`, 0, -1);

      if (!myPids.length) {
        res.json(ok({ records: [], page, pageCount: 1 }));
        return;
      }

      // 收集所有对用户帖子的点赞
      const records = [];
      for (const pid of myPids) {
        const upvoters = await db.getSetMembers(`pid:${pid}:upvote`);
        if (upvoters.length > 0) {
          const postData = await posts.getPostFields(pid, ['pid', 'tid', 'content', 'timestamp']);
          for (const voterUid of upvoters) {
            if (parseInt(voterUid, 10) !== req.uid) {
              const voterData = await user.getUserFields(voterUid, ['uid', 'username', 'picture']);
              records.push({
                type: 'upvote',
                post: postData,
                fromUser: voterData,
                timestamp: postData.timestamp,
              });
            }
          }
        }
      }

      // 按时间排序并分页
      records.sort((a, b) => b.timestamp - a.timestamp);
      const paged = records.slice(start, stop + 1);

      res.json(ok({
        records: paged,
        page,
        pageCount: Math.ceil(records.length / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取别人对我的评论（被评论记录）
   * GET /api/moti/user/received-replies
   */
  api.get('/user/received-replies', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);

      // 获取用户发表的帖子
      const myPids = await db.getSortedSetRevRange(`uid:${req.uid}:posts`, 0, -1);

      if (!myPids.length) {
        res.json(ok({ records: [], page, pageCount: 1 }));
        return;
      }

      // 查找回复了用户帖子的评论
      const records = [];
      for (const pid of myPids) {
        // 获取回复这个帖子的帖子
        const replyPids = await db.getSortedSetRevRange(`pid:${pid}:replies`, 0, 20);
        for (const replyPid of replyPids) {
          const replyData = await posts.getPostData(replyPid);
          if (replyData && parseInt(replyData.uid, 10) !== req.uid) {
            const replyUser = await user.getUserFields(replyData.uid, ['uid', 'username', 'picture']);
            records.push({
              type: 'reply',
              post: replyData,
              fromUser: replyUser,
              timestamp: replyData.timestamp,
            });
          }
        }
      }

      // 按时间排序并分页
      records.sort((a, b) => b.timestamp - a.timestamp);
      const paged = records.slice(start, stop + 1);

      res.json(ok({
        records: paged,
        page,
        pageCount: Math.ceil(records.length / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取用户积分信息
   * GET /api/moti/user/points
   */
  api.get('/user/points', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const points = await db.getObjectField(`user:${req.uid}`, 'points') || 0;
      const reputation = await user.getUserField(req.uid, 'reputation') || 0;

      res.json(ok({
        points: parseInt(points, 10) || 0,
        reputation: parseInt(reputation, 10) || 0,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取互动统计概览
   * GET /api/moti/user/stats
   */
  api.get('/user/stats', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const [
        bookmarkCount,
        upvoteCount,
        postCount,
        noteCount,
      ] = await Promise.all([
        db.sortedSetCard(`uid:${req.uid}:bookmarks`),
        db.sortedSetCard(`uid:${req.uid}:upvote`),
        db.sortedSetCard(`uid:${req.uid}:posts`),
        db.sortedSetCard(`uid:${req.uid}:notes`),
      ]);

      const userData = await user.getUserFields(req.uid, ['reputation', 'postcount', 'topiccount']);

      res.json(ok({
        bookmarks: bookmarkCount || 0,
        upvotes: upvoteCount || 0,
        posts: postCount || 0,
        notes: noteCount || 0,
        reputation: userData.reputation || 0,
        postcount: userData.postcount || 0,
        topiccount: userData.topiccount || 0,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 支付相关路由 ============

  /**
   * 获取 VIP 商品列表
   * GET /api/moti/products
   */
  api.get('/products', (req, res) => {
    try {
      const products = payment.getProducts();
      res.json(ok({ products }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取 VIP 状态
   * GET /api/moti/vip/status
   */
  api.get('/vip/status', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const status = await payment.checkVipStatus(req.uid);
      res.json(ok(status));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 创建支付订单
   * POST /api/moti/order/create
   * Body: { productId: 'vip_month' | 'vip_quarter' | 'vip_year' }
   */
  api.post('/order/create', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { productId } = req.body || {};

      if (!productId) {
        res.status(400).json(err('productId-required'));
        return;
      }

      // 获取用户的 openid
      const wechatBinding = await wechat.getWechatBinding(req.uid);
      if (!wechatBinding || !wechatBinding.openid) {
        res.status(400).json(err('wechat-bindding-required'));
        return;
      }

      // 创建订单
      const order = await payment.createOrder(req.uid, productId, wechatBinding.openid);

      // 调用微信支付
      const payResult = await payment.createWechatPayOrder(order);

      res.json(ok({
        orderNo: order.orderNo,
        amount: order.amount,
        productName: order.productName,
        payParams: payResult.payParams,
      }));
    } catch (e) {
      console.error('[moti-api] create order error:', e);
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 查询订单状态
   * GET /api/moti/order/:orderNo
   */
  api.get('/order/:orderNo', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const orderNo = String(req.params.orderNo);
      const order = await payment.getOrder(orderNo);

      if (!order) {
        res.status(404).json(err('order-not-found'));
        return;
      }

      // 只能查看自己的订单
      if (parseInt(order.uid, 10) !== req.uid) {
        res.status(403).json(err('not-allowed'));
        return;
      }

      res.json(ok(order));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取我的订单列表
   * GET /api/moti/orders
   */
  api.get('/orders', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);
      const orders = await payment.getUserOrders(req.uid, start, stop);
      const total = await db.sortedSetCard(`uid:${req.uid}:orders`);

      res.json(ok({
        orders,
        page,
        pageCount: Math.ceil(total / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取付费记录（已支付的订单）
   * GET /api/moti/payments
   */
  api.get('/payments', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);
      const payments = await payment.getPaymentHistory(req.uid, start, stop);

      res.json(ok({
        payments,
        page,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 微信支付回调
   * POST /api/moti/payment/callback
   * 注意：此接口由微信服务器调用，不需要认证
   */
  api.post('/payment/callback', express.raw({ type: 'application/json' }), async (req, res) => {
    try {
      const headers = req.headers;
      const body = req.body.toString('utf8');

      const result = await payment.handlePaymentCallback(headers, body);

      if (result.success) {
        res.status(200).json({ code: 'SUCCESS', message: '成功' });
      } else {
        res.status(200).json({ code: 'FAIL', message: result.message });
      }
    } catch (e) {
      console.error('[moti-api] payment callback error:', e);
      res.status(500).json({ code: 'FAIL', message: e.message });
    }
  });

  // ============ 提问发帖路由 ============

  /**
   * 检查是否有发帖权限（VIP用户）
   */
  async function checkPostPermission(req, res, next) {
    const uid = parseInt(req.uid, 10) || 0;
    if (uid <= 0) {
      res.status(401).json(err('not-logged-in'));
      return;
    }

    const vipExpire = await db.getObjectField(`user:${uid}`, 'vip_expire_time');
    if (!vipExpire || parseInt(vipExpire, 10) <= Date.now()) {
      res.status(403).json(err('vip-required'));
      return;
    }
    next();
  }

  /**
   * 发布提问帖子
   * POST /api/moti/topic/create
   * Body: { title, content, cid, tags? }
   * 需要 VIP 权限
   */
  api.post('/topic/create', [authenticateMotiToken, requireLogin, checkPostPermission], async (req, res) => {
    try {
      const { title, content, cid, tags } = req.body || {};

      if (!title || !title.trim()) {
        res.status(400).json(err('title-required'));
        return;
      }
      if (!content || !content.trim()) {
        res.status(400).json(err('content-required'));
        return;
      }
      if (!cid) {
        res.status(400).json(err('cid-required'));
        return;
      }

      const topicData = await topics.post({
        uid: req.uid,
        cid: parseInt(cid, 10),
        title: String(title).trim(),
        content: String(content).trim(),
        tags: Array.isArray(tags) ? tags : (tags ? String(tags).split(',').map(t => t.trim()).filter(Boolean) : []),
      });

      res.json(ok({
        tid: topicData.tid,
        topic: topicData,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取我的提问列表
   * GET /api/moti/user/topics
   */
  api.get('/user/topics', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);
      const tids = await db.getSortedSetRevRange(`uid:${req.uid}:topics`, start, stop);
      const total = await db.sortedSetCard(`uid:${req.uid}:topics`);

      if (!tids.length) {
        res.json(ok({ topics: [], page, pageCount: 1 }));
        return;
      }

      const topicsData = await topics.getTopics(tids, req.uid);

      res.json(ok({
        topics: topicsData.filter(Boolean),
        page,
        pageCount: Math.ceil(total / limit) || 1,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 投票功能路由 ============

  /**
   * 对帖子投票（点赞/踩）
   * POST /api/moti/post/:pid/vote
   * Body: { delta: 1 | -1 | 0 } (1=赞, -1=踩, 0=取消)
   */
  api.post('/post/:pid/vote', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const pid = parseInt(req.params.pid, 10);
      const delta = parseInt(req.body.delta, 10);

      let result;
      if (delta === 1) {
        result = await posts.upvote(pid, req.uid);
      } else if (delta === -1) {
        result = await posts.downvote(pid, req.uid);
      } else {
        result = await posts.unvote(pid, req.uid);
      }

      // 给被点赞用户加积分
      if (delta === 1 && result.upvote) {
        const postUid = await posts.getPostField(pid, 'uid');
        if (postUid && parseInt(postUid, 10) !== req.uid) {
          await db.incrObjectField(`user:${postUid}`, 'points');
        }
      }

      res.json(ok({
        upvoted: result.upvote || false,
        downvoted: result.downvote || false,
        votes: result.post ? result.post.votes : 0,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 对主题投票
   * POST /api/moti/topic/:tid/vote
   * Body: { delta: 1 | -1 | 0 }
   */
  api.post('/topic/:tid/vote', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const tid = parseInt(req.params.tid, 10);
      const delta = parseInt(req.body.delta, 10);

      // 获取主帖 pid
      const mainPid = await topics.getTopicField(tid, 'mainPid');

      let result;
      if (delta === 1) {
        result = await posts.upvote(mainPid, req.uid);
      } else if (delta === -1) {
        result = await posts.downvote(mainPid, req.uid);
      } else {
        result = await posts.unvote(mainPid, req.uid);
      }

      res.json(ok({
        upvoted: result.upvote || false,
        downvoted: result.downvote || false,
        votes: result.post ? result.post.votes : 0,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 举报功能路由 ============

  /**
   * 举报帖子/评论
   * POST /api/moti/report
   * Body: { type: 'post', id: pid, reason: string }
   */
  api.post('/report', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { type, id, reason } = req.body || {};

      if (!type || !id) {
        res.status(400).json(err('type-and-id-required'));
        return;
      }
      if (!reason || !reason.trim()) {
        res.status(400).json(err('reason-required'));
        return;
      }

      const reportId = crypto.randomUUID();
      const now = Date.now();

      const report = {
        reportId,
        type,
        targetId: parseInt(id, 10),
        reason: String(reason).trim().substring(0, 500),
        reporterUid: req.uid,
        status: 'pending', // pending, reviewed, resolved, rejected
        createTime: now,
        updateTime: now,
      };

      // 获取被举报内容的信息
      if (type === 'post') {
        const postData = await posts.getPostFields(id, ['uid', 'tid', 'content']);
        report.targetUid = postData.uid;
        report.targetTid = postData.tid;
        report.targetContent = String(postData.content || '').substring(0, 200);
      }

      // 保存举报
      await db.setObject(`report:${reportId}`, report);
      await db.sortedSetAdd('reports:pending', now, reportId);
      await db.sortedSetAdd(`uid:${req.uid}:reports`, now, reportId);

      res.json(ok({ reportId, message: 'report-submitted' }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 获取我的举报记录
   * GET /api/moti/user/reports
   */
  api.get('/user/reports', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);
      const reportIds = await db.getSortedSetRevRange(`uid:${req.uid}:reports`, start, stop);

      if (!reportIds.length) {
        res.json(ok({ reports: [], page, pageCount: 1 }));
        return;
      }

      const reports = await Promise.all(reportIds.map(id => db.getObject(`report:${id}`)));

      res.json(ok({
        reports: reports.filter(Boolean),
        page,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  // ============ 积分系统 ============

  /**
   * 积分规则配置
   */
  const POINTS_RULES = {
    CREATE_TOPIC: 10,      // 发帖
    CREATE_REPLY: 5,       // 评论/回复
    RECEIVE_UPVOTE: 2,     // 被点赞
    DAILY_CHECKIN: 5,      // 每日签到
    CREATE_NOTE: 3,        // 划线笔记
    SHARE_CONTENT: 2,      // 分享内容
    INVITE_USER: 20,       // 邀请好友注册
  };

  /**
   * 增加用户积分
   * @param {number} uid - 用户ID
   * @param {number} points - 积分数量
   * @param {string} reason - 积分来源
   * @param {object} extra - 额外信息
   */
  async function addPoints(uid, points, reason, extra = {}) {
    if (!uid || points <= 0) return;

    const now = Date.now();

    // 增加积分
    await db.incrObjectFieldBy(`user:${uid}`, 'moti_points', points);

    // 获取新的总积分
    const totalPoints = await db.getObjectField(`user:${uid}`, 'moti_points') || 0;

    // 更新排行榜
    await db.sortedSetAdd('moti:points:ranking', totalPoints, uid);

    // 记录积分日志
    const logId = `${uid}:${now}:${Math.random().toString(36).substr(2, 6)}`;
    const logData = {
      logId,
      uid,
      points,
      reason,
      createTime: now,
      ...extra,
    };
    await db.setObject(`moti:points:log:${logId}`, logData);
    await db.sortedSetAdd(`user:${uid}:points:log`, now, logId);

    return totalPoints;
  }

  /**
   * 获取用户积分
   * GET /api/moti/points
   */
  api.get('/points', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const points = await db.getObjectField(`user:${req.uid}`, 'moti_points') || 0;
      const rank = await db.sortedSetRevRank('moti:points:ranking', req.uid);

      // 获取今日是否已签到
      const today = new Date().toISOString().slice(0, 10);
      const lastCheckin = await db.getObjectField(`user:${req.uid}`, 'moti_last_checkin');
      const checkedInToday = lastCheckin === today;

      // 获取连续签到天数
      const checkinStreak = parseInt(await db.getObjectField(`user:${req.uid}`, 'moti_checkin_streak') || 0, 10);

      res.json(ok({
        points: parseInt(points, 10),
        rank: rank !== null ? rank + 1 : null,
        checkedInToday,
        checkinStreak,
        rules: POINTS_RULES,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 每日签到
   * POST /api/moti/points/checkin
   */
  api.post('/points/checkin', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const today = new Date().toISOString().slice(0, 10);
      const lastCheckin = await db.getObjectField(`user:${req.uid}`, 'moti_last_checkin');

      if (lastCheckin === today) {
        res.status(400).json(err('already-checked-in-today'));
        return;
      }

      // 计算连续签到
      let streak = parseInt(await db.getObjectField(`user:${req.uid}`, 'moti_checkin_streak') || 0, 10);
      const yesterday = new Date(Date.now() - 86400000).toISOString().slice(0, 10);

      if (lastCheckin === yesterday) {
        streak += 1;
      } else {
        streak = 1;
      }

      // 连续签到奖励: 7天+5, 30天+10
      let bonusPoints = 0;
      if (streak === 7) bonusPoints = 5;
      if (streak === 30) bonusPoints = 10;

      const basePoints = POINTS_RULES.DAILY_CHECKIN;
      const totalEarned = basePoints + bonusPoints;

      // 更新签到记录
      await db.setObject(`user:${req.uid}`, {
        moti_last_checkin: today,
        moti_checkin_streak: streak,
      });

      // 增加积分
      const totalPoints = await addPoints(req.uid, totalEarned, 'checkin', { streak });

      res.json(ok({
        points: totalEarned,
        streak,
        bonusPoints,
        totalPoints,
        message: bonusPoints > 0 ? `连续签到${streak}天，额外奖励${bonusPoints}积分！` : '签到成功！',
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 积分排行榜
   * GET /api/moti/points/ranking
   */
  api.get('/points/ranking', authenticateMotiToken, async (req, res) => {
    try {
      const limit = Math.min(parseInt(req.query.limit, 10) || 50, 100);

      // 获取排行榜
      const uidsWithScores = await db.getSortedSetRevRangeWithScores('moti:points:ranking', 0, limit - 1);

      if (!uidsWithScores.length) {
        res.json(ok({ ranking: [], myRank: null }));
        return;
      }

      // 获取用户信息
      const uids = uidsWithScores.map(item => item.value);
      const users = await user.getUsersFields(uids, ['uid', 'username', 'picture']);

      const ranking = uidsWithScores.map((item, index) => {
        const userData = users.find(u => String(u.uid) === String(item.value)) || {};
        return {
          rank: index + 1,
          uid: parseInt(item.value, 10),
          username: userData.username || '用户',
          picture: userData.picture || '',
          points: parseInt(item.score, 10),
        };
      });

      // 获取当前用户排名
      let myRank = null;
      if (req.loggedIn && req.uid) {
        const rank = await db.sortedSetRevRank('moti:points:ranking', req.uid);
        if (rank !== null) {
          const myPoints = await db.getObjectField(`user:${req.uid}`, 'moti_points') || 0;
          myRank = {
            rank: rank + 1,
            points: parseInt(myPoints, 10),
          };
        }
      }

      res.json(ok({ ranking, myRank }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 积分历史记录
   * GET /api/moti/points/history
   */
  api.get('/points/history', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { page, limit, start, stop } = await getPagination(req);
      const logIds = await db.getSortedSetRevRange(`user:${req.uid}:points:log`, start, stop);

      if (!logIds.length) {
        res.json(ok({ logs: [], page, pageCount: 1 }));
        return;
      }

      const logs = await Promise.all(logIds.map(id => db.getObject(`moti:points:log:${id}`)));

      // 转换原因为可读文本
      const reasonMap = {
        checkin: '每日签到',
        create_topic: '发布帖子',
        create_reply: '发表评论',
        receive_upvote: '收到点赞',
        create_note: '创建笔记',
        share_content: '分享内容',
        invite_user: '邀请好友',
      };

      const formattedLogs = logs.filter(Boolean).map(log => ({
        ...log,
        reasonText: reasonMap[log.reason] || log.reason,
      }));

      res.json(ok({
        logs: formattedLogs,
        page,
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });

  /**
   * 分享内容获取积分
   * POST /api/moti/points/share
   * Body: { type: 'topic'|'post', id: number }
   */
  api.post('/points/share', [authenticateMotiToken, requireLogin], async (req, res) => {
    try {
      const { type, id } = req.body || {};
      if (!type || !id) {
        res.status(400).json(err('type-and-id-required'));
        return;
      }

      // 每个内容每天只能获得一次分享积分
      const today = new Date().toISOString().slice(0, 10);
      const shareKey = `user:${req.uid}:share:${type}:${id}:${today}`;
      const alreadyShared = await db.get(shareKey);

      if (alreadyShared) {
        res.json(ok({ points: 0, message: '今日已分享过该内容' }));
        return;
      }

      // 标记已分享
      await db.set(shareKey, '1');
      await db.pexpire(shareKey, 86400000); // 24小时后过期

      // 增加积分
      const totalPoints = await addPoints(req.uid, POINTS_RULES.SHARE_CONTENT, 'share_content', { type, targetId: id });

      res.json(ok({
        points: POINTS_RULES.SHARE_CONTENT,
        totalPoints,
        message: '分享成功！',
      }));
    } catch (e) {
      res.status(400).json(err(e.message));
    }
  });
};

// ============ 积分钩子函数（在发帖、回复、点赞时自动增加积分）============

/**
 * 发帖时增加积分
 */
plugin.onTopicCreate = async function (data) {
  const POINTS_RULES = {
    CREATE_TOPIC: 10,
  };

  if (data.topic && data.topic.uid) {
    const now = Date.now();
    const uid = data.topic.uid;
    const points = POINTS_RULES.CREATE_TOPIC;

    await db.incrObjectFieldBy(`user:${uid}`, 'moti_points', points);
    const totalPoints = await db.getObjectField(`user:${uid}`, 'moti_points') || 0;
    await db.sortedSetAdd('moti:points:ranking', totalPoints, uid);

    const logId = `${uid}:${now}:${Math.random().toString(36).substr(2, 6)}`;
    await db.setObject(`moti:points:log:${logId}`, {
      logId, uid, points, reason: 'create_topic', createTime: now, tid: data.topic.tid,
    });
    await db.sortedSetAdd(`user:${uid}:points:log`, now, logId);
  }
  return data;
};

/**
 * 回复时增加积分
 */
plugin.onPostCreate = async function (data) {
  const POINTS_RULES = {
    CREATE_REPLY: 5,
  };

  // 只有回复才给积分，主帖已在 onTopicCreate 中处理
  if (data.post && data.post.uid && !data.post.isMain) {
    const now = Date.now();
    const uid = data.post.uid;
    const points = POINTS_RULES.CREATE_REPLY;

    await db.incrObjectFieldBy(`user:${uid}`, 'moti_points', points);
    const totalPoints = await db.getObjectField(`user:${uid}`, 'moti_points') || 0;
    await db.sortedSetAdd('moti:points:ranking', totalPoints, uid);

    const logId = `${uid}:${now}:${Math.random().toString(36).substr(2, 6)}`;
    await db.setObject(`moti:points:log:${logId}`, {
      logId, uid, points, reason: 'create_reply', createTime: now, pid: data.post.pid,
    });
    await db.sortedSetAdd(`user:${uid}:points:log`, now, logId);
  }
  return data;
};

/**
 * 被点赞时增加积分
 */
plugin.onPostUpvote = async function (data) {
  const POINTS_RULES = {
    RECEIVE_UPVOTE: 2,
  };

  // 给被点赞的帖子作者增加积分
  if (data.post && data.post.uid && data.current && data.current.uid !== data.post.uid) {
    const now = Date.now();
    const uid = data.post.uid;
    const points = POINTS_RULES.RECEIVE_UPVOTE;

    await db.incrObjectFieldBy(`user:${uid}`, 'moti_points', points);
    const totalPoints = await db.getObjectField(`user:${uid}`, 'moti_points') || 0;
    await db.sortedSetAdd('moti:points:ranking', totalPoints, uid);

    const logId = `${uid}:${now}:${Math.random().toString(36).substr(2, 6)}`;
    await db.setObject(`moti:points:log:${logId}`, {
      logId, uid, points, reason: 'receive_upvote', createTime: now, pid: data.post.pid, fromUid: data.current.uid,
    });
    await db.sortedSetAdd(`user:${uid}:points:log`, now, logId);
  }
  return data;
};

module.exports = plugin;
