// nodebb-plugin-moti-api/library.js
'use strict';

const winston = require('winston');
const nconf = require('nconf');

// 引入 NodeBB 核心模块
const database = require.main.require('./src/database');
const user = require.main.require('./src/user');
const posts = require.main.require('./src/posts');
const topics = require.main.require('./src/topics');

const plugin = {};

// ============================================
// 插件初始化
// ============================================

plugin.init = async function (params) {
  const { router, middleware } = params;

  winston.info('[plugin/moti-api] 莫提之地API插件正在初始化...');

  // 注册API路由
  registerRoutes(router, middleware);

  winston.info('[plugin/moti-api] 莫提之地API插件初始化完成');
};

// ============================================
// 路由注册
// ============================================

function registerRoutes(router, middleware) {
  // 测试路由（不需要中间件）
  router.get('/api/moti/test', (req, res) => {
    res.json({
      success: true,
      message: '莫提之地API插件运行正常',
      version: '1.0.0',
      timestamp: Date.now()
    });
  });

  // ==================== 用户相关API ====================

  // 获取用户信息（需要认证）
  router.get('/api/moti/user/profile',
    middleware.ensureLoggedIn,
    getUserProfile
  );

  // 微信登录（待实现）
  router.post('/api/moti/auth/wechat', wechatLogin);

  // 绑定手机号（需要认证）
  router.post('/api/moti/auth/phone', middleware.ensureLoggedIn, bindPhone);

  // ==================== 内容相关API ====================

  // 获取帖子列表
  router.get('/api/moti/posts', getPosts);

  // 获取帖子详情（带权限控制）
  router.get('/api/moti/post/:pid', getPost);

  // ==================== 互动相关API ====================

  // 点赞（需要认证）
  router.post('/api/moti/post/:pid/upvote', middleware.ensureLoggedIn, upvotePost);

  // 收藏（需要认证）
  router.post('/api/moti/post/:pid/bookmark', middleware.ensureLoggedIn, bookmarkPost);

  // ==================== 笔记相关API ====================

  // 保存笔记（划线，需要认证）
  router.post('/api/moti/notes', middleware.ensureLoggedIn, saveNote);

  // 获取用户的所有笔记（需要认证）
  router.get('/api/moti/notes', middleware.ensureLoggedIn, getUserNotes);

  // 获取指定帖子的笔记（需要认证）
  router.get('/api/moti/post/:pid/notes', middleware.ensureLoggedIn, getPostNotes);

  // 更新笔记（需要认证）
  router.put('/api/moti/notes/:noteId', middleware.ensureLoggedIn, updateNote);

  // 删除笔记（需要认证）
  router.delete('/api/moti/notes/:noteId', middleware.ensureLoggedIn, deleteNote);

  winston.info('[plugin/moti-api] API路由注册完成');
}

// ============================================
// API 处理函数
// ============================================

/**
 * 获取用户信息
 */
async function getUserProfile(req, res) {
  try {
    // const user = require('../../../src/user'); // 使用顶部引入的 user 模块
    const uid = req.uid;

    if (!uid) {
      return res.status(401).json({ error: '未登录' });
    }

    const userData = await user.getUserData(uid);

    // 获取VIP状态和积分
    const db = database;
    const vipExpire = await db.getObjectField(`user:${uid}`, 'vip_expire_time');
    const points = await db.getObjectField(`user:${uid}`, 'points') || 0;
    const dailyReads = await db.getObjectField(`user:${uid}`, 'daily_read_count') || 0;

    res.json({
      success: true,
      data: {
        uid: userData.uid,
        username: userData.username,
        email: userData.email,
        picture: userData.picture,
        isVip: vipExpire && vipExpire > Date.now(),
        vipExpire: vipExpire || null,
        points: parseInt(points, 10),
        dailyReads: parseInt(dailyReads, 10),
        reputation: userData.reputation,
        postcount: userData.postcount,
        joindate: userData.joindate
      }
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 获取用户信息失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 微信登录
 */
async function wechatLogin(req, res) {
  try {
    const { code } = req.body;

    if (!code) {
      return res.status(400).json({ error: '缺少微信code参数' });
    }

    // TODO: 调用微信API获取openid
    // TODO: 查询或创建用户
    // TODO: 生成token

    res.json({
      success: true,
      message: '微信登录功能开发中',
      data: {
        token: 'mock_token',
        uid: 1
      }
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 微信登录失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 绑定手机号
 */
async function bindPhone(req, res) {
  try {
    const { phone, code } = req.body;
    const uid = req.uid;

    if (!phone || !code) {
      return res.status(400).json({ error: '缺少必要参数' });
    }

    // TODO: 验证短信验证码
    // TODO: 绑定手机号到用户

    const db = database;
    await db.setObjectField(`user:${uid}`, 'phone', phone);

    res.json({
      success: true,
      message: '手机号绑定成功'
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 绑定手机号失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 获取帖子列表
 */
async function getPosts(req, res) {
  try {
    // const topics = require('../../../src/topics'); // 使用顶部引入的 topics 模块
    const page = parseInt(req.query.page, 10) || 1;
    const sort = req.query.sort || 'recent'; // recent, popular, mostbookmarked
    const limit = 20;

    let method;
    switch (sort) {
      case 'popular':
        method = 'getPopular';
        break;
      case 'mostbookmarked':
        method = 'getTopics';
        break;
      default:
        method = 'getLatestTopics';
    }

    const result = await topics[method](req.uid, page - 1, limit);

    res.json({
      success: true,
      data: {
        topics: result.topics || [],
        page: page,
        pageCount: result.pageCount || 1
      }
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 获取帖子列表失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 获取帖子详情（带阅读权限检查）
 */
async function getPost(req, res) {
  try {
    const pid = parseInt(req.params.pid, 10);
    const uid = req.uid || 0;

    // 检查阅读权限
    const canRead = await checkUserReadPermission(uid);
    if (!canRead) {
      return res.status(403).json({
        error: '今日免费阅读次数已用完',
        code: 'DAILY_LIMIT_EXCEEDED',
        needVip: true
      });
    }

    // const posts = require('../../../src/posts'); // 使用顶部引入的 posts 模块
    const postData = await posts.getPostData(pid);

    if (!postData) {
      return res.status(404).json({ error: '帖子不存在' });
    }

    // 记录阅读
    if (uid > 0) {
      await recordRead(uid);
    }

    res.json({
      success: true,
      data: postData
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 获取帖子详情失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 点赞帖子
 */
async function upvotePost(req, res) {
  try {
    const pid = parseInt(req.params.pid, 10);
    const uid = req.uid;

    // const posts = require('../../../src/posts'); // 使用顶部引入的 posts 模块
    await posts.upvote(pid, uid);

    res.json({
      success: true,
      message: '点赞成功'
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 点赞失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 收藏帖子
 */
async function bookmarkPost(req, res) {
  try {
    const pid = parseInt(req.params.pid, 10);
    const uid = req.uid;

    // const posts = require('../../../src/posts'); // 使用顶部引入的 posts 模块
    await posts.bookmark(pid, uid);

    res.json({
      success: true,
      message: '收藏成功'
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 收藏失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 保存笔记（划线）
 */
async function saveNote(req, res) {
  try {
    const uid = req.uid;
    const { pid, content, note, startOffset, endOffset, color } = req.body;

    if (!pid || !content || startOffset === undefined || endOffset === undefined) {
      return res.status(400).json({ error: '缺少必要参数' });
    }

    const db = database;
    const noteId = `note_${uid}_${Date.now()}`;
    const timestamp = Date.now();

    // 保存笔记数据
    await db.setObject(`note:${noteId}`, {
      noteId: noteId,
      uid: uid,
      pid: parseInt(pid, 10),
      content: content,
      note: note || '',
      startOffset: parseInt(startOffset, 10),
      endOffset: parseInt(endOffset, 10),
      color: color || '#ffeb3b',
      timestamp: timestamp
    });

    // 添加到用户笔记索引
    await db.sortedSetAdd(`user:${uid}:notes`, timestamp, noteId);

    // 添加到帖子-用户笔记索引
    await db.sortedSetAdd(`post:${pid}:user:${uid}:notes`, timestamp, noteId);

    res.json({
      success: true,
      message: '笔记保存成功',
      data: { noteId: noteId }
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 保存笔记失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 获取用户的所有笔记
 */
async function getUserNotes(req, res) {
  try {
    const uid = req.uid;
    const page = parseInt(req.query.page, 10) || 1;
    const limit = parseInt(req.query.limit, 10) || 20;
    const start = (page - 1) * limit;
    const stop = start + limit - 1;

    const db = database;

    // 获取笔记ID列表（按时间倒序）
    const noteIds = await db.getSortedSetRevRange(`user:${uid}:notes`, start, stop);

    // 获取笔记详情
    const notes = [];
    for (const noteId of noteIds) {
      const noteData = await db.getObject(`note:${noteId}`);
      if (noteData) {
        notes.push(noteData);
      }
    }

    // 获取总数
    const total = await db.sortedSetCard(`user:${uid}:notes`);

    res.json({
      success: true,
      data: {
        notes: notes,
        page: page,
        limit: limit,
        total: total,
        pageCount: Math.ceil(total / limit)
      }
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 获取笔记列表失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 获取指定帖子的笔记
 */
async function getPostNotes(req, res) {
  try {
    const uid = req.uid;
    const pid = parseInt(req.params.pid, 10);

    const db = database;

    // 获取该帖子的所有笔记ID（按时间倒序）
    const noteIds = await db.getSortedSetRevRange(`post:${pid}:user:${uid}:notes`, 0, -1);

    // 获取笔记详情
    const notes = [];
    for (const noteId of noteIds) {
      const noteData = await db.getObject(`note:${noteId}`);
      if (noteData) {
        notes.push(noteData);
      }
    }

    res.json({
      success: true,
      data: {
        pid: pid,
        notes: notes,
        total: notes.length
      }
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 获取帖子笔记失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 更新笔记
 */
async function updateNote(req, res) {
  try {
    const uid = req.uid;
    const noteId = req.params.noteId;
    const { note, color } = req.body;

    const db = database;

    // 验证笔记所有权
    const noteData = await db.getObject(`note:${noteId}`);
    if (!noteData) {
      return res.status(404).json({ error: '笔记不存在' });
    }

    if (parseInt(noteData.uid, 10) !== uid) {
      return res.status(403).json({ error: '无权限修改此笔记' });
    }

    // 更新笔记
    const updates = {};
    if (note !== undefined) {
      updates.note = note;
    }
    if (color !== undefined) {
      updates.color = color;
    }

    await db.setObject(`note:${noteId}`, updates);

    res.json({
      success: true,
      message: '笔记更新成功'
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 更新笔记失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

/**
 * 删除笔记
 */
async function deleteNote(req, res) {
  try {
    const uid = req.uid;
    const noteId = req.params.noteId;

    const db = database;

    // 验证笔记所有权
    const noteData = await db.getObject(`note:${noteId}`);
    if (!noteData) {
      return res.status(404).json({ error: '笔记不存在' });
    }

    if (parseInt(noteData.uid, 10) !== uid) {
      return res.status(403).json({ error: '无权限删除此笔记' });
    }

    // 从索引中移除
    await db.sortedSetRemove(`user:${uid}:notes`, noteId);
    await db.sortedSetRemove(`post:${noteData.pid}:user:${uid}:notes`, noteId);

    // 删除笔记数据
    await db.delete(`note:${noteId}`);

    res.json({
      success: true,
      message: '笔记删除成功'
    });
  } catch (error) {
    winston.error('[plugin/moti-api] 删除笔记失败:', error);
    res.status(500).json({ error: '服务器错误' });
  }
}

// ============================================
// 权限控制辅助函数
// ============================================

/**
 * 检查用户是否有阅读权限
 */
async function checkUserReadPermission(uid) {
  if (uid === 0) {
    // 未登录用户不允许阅读
    return false;
  }

  const db = database;

  // 检查是否VIP
  const vipExpire = await db.getObjectField(`user:${uid}`, 'vip_expire_time');
  if (vipExpire && vipExpire > Date.now()) {
    return true; // VIP用户无限制
  }

  // 检查今日阅读次数
  const today = new Date().toDateString();
  const lastReadDate = await db.getObjectField(`user:${uid}`, 'last_read_date');
  const dailyReads = parseInt(await db.getObjectField(`user:${uid}`, 'daily_read_count') || 0, 10);

  if (lastReadDate !== today) {
    // 新的一天，重置计数
    await db.setObject(`user:${uid}`, {
      last_read_date: today,
      daily_read_count: 0
    });
    return true;
  }

  // 检查是否超过每日限制
  return dailyReads < 1; // 每日限读1篇
}

/**
 * 记录阅读
 */
async function recordRead(uid) {
  const db = database;

  // 检查是否VIP
  const vipExpire = await db.getObjectField(`user:${uid}`, 'vip_expire_time');
  if (vipExpire && vipExpire > Date.now()) {
    return; // VIP用户不计数
  }

  await db.incrObjectField(`user:${uid}`, 'daily_read_count');
}

// ============================================
// Hooks
// ============================================

/**
 * 用户创建时初始化扩展字段
 */
plugin.onUserCreate = async function (data) {
  const { user } = data;
  const db = database;

  await db.setObject(`user:${user.uid}`, {
    vip_expire_time: 0,
    points: 0,
    daily_read_count: 0,
    last_read_date: new Date().toDateString()
  });

  return data;
};

/**
 * 获取帖子时检查权限（hook）
 */
plugin.checkReadPermission = async function (data) {
  // 这个hook可以在获取帖子时自动检查权限
  return data;
};

/**
 * 添加客户端脚本
 */
plugin.addScripts = async function (data) {
  if (!data.scripts) {
    data.scripts = [];
  }
  data.scripts.push('/assets/plugins/nodebb-plugin-moti-api/static/lib/highlight.js');
  return data;
};

module.exports = plugin;
